core/src/main/scala-3/medeia/generic/GenericProductDecoder.scala

```diff
@@ -33,30 +35,29 @@ object GenericProductDecoder {
       labelling: Labelling[A],
       options: GenericDerivationOptions[A]
   ): EitherNec[BsonDecoderError, A] = {
-    val (acc, result) = inst.unfold[Accumulator](Accumulator(labelling.elemLabels, Chain.empty))(
+    val (acc, result) = inst.unfold[Accumulator](Accumulator(labelling.elemLabels, None))(
       [t] =>
-        (acc: Accumulator, rt: BsonDecoder[t]) => {
-          val label = options.transformKeys(acc.labels.head)
+        (acc: Accumulator, rt: BsonDecoder[t]) =>
+          val label =
+            acc.labels.headOption
+              .map(options.transformKeys)
+              .getOrElse(throw new RuntimeException("Bug in derivation logic, accumulator has no labels!"))
+
           val decoded = Option(bsonDocument.get(label)) match {
             case Some(headField) => rt.decode(headField).leftMap(_.map(_.push(Attr(label))))
             case None            => rt.defaultValue.toRight(NonEmptyChain(KeyNotFound(label)))
           }
+
           decoded match {
-            case Left(error) =>
-              (acc.next(error), None)
-            case Right(value) =>
-              (acc.next(), Some(value))
-          }
-      }
+            case Left(es)     => (acc.addErrors(es), Some(null.asInstanceOf[t])) // we have to avoid stopping the unfold via None
```

Comment:
Ugly hack following: If we return `None` here, the unfold will stop it's work and we don't get all of the errors.  Instead we smuggle `null`s through all the way and below check if there were any errors.  Awkward, but I did not find any good solution to this.  

Maybe @joroKr21 has an idea?  
We need to be able to continue the `unfold` but looking at the public API of https://github.com/typelevel/shapeless-3 I don't see a way to do that.
Anyway, this could be interesting for you in case https://github.com/moia-oss/scynamo get's a Scala 3 update :wink:

Reply:
A bit out of context here - why do you need to continue unfolding? You will not get a valid instance back this way (one of the fields will be null). If you want to accumulate errors perhaps the right abstraction is `traverse`?

Reply:
Hmm maybe there is something missing in the shapeless API. Can you think of a type signature that would work for this use case?

Reply:
I will merge this branch onto my feature branch and then that branch onto main. If there is a coming change in shapeless which makes our code better, we can open another pull request for it I think :) Please feel free to discuss it here :)

Thanks for your review @markus1189 @joroKr21, I really appreciate it.